crates/core/tedge_actors/src/channels.rs

```diff
@@ -3,128 +3,118 @@ use crate::ChannelError;
+        Ok(self.as_mut().send(message.into()).await?)
+    }
+}
+
+#[async_trait]
+impl<M: Message, N: Message + Into<M>> Sender<N> for Box<dyn Sender<M>> {
+    async fn send(&mut self, message: N) -> Result<(), ChannelError> {
+        Ok(self.as_mut().send(message.into()).await?)
     }
+}
 
+#[async_trait]
+impl<M: Message, N: Message + Into<M>> CloneSender<N> for DynSender<M> {
     fn sender_clone(&self) -> DynSender<N> {
-        Box::new(self.clone())
+        Box::new(self.as_ref().sender_clone())
     }
 
-    fn close_sender(&mut self) {
-        self.close_channel();
+    fn sender(&self) -> Box<dyn Sender<N>> {
+        Box::new(self.as_ref().sender())
     }
 }
 
-/// An `mpsc::UnboundedSender<M>` is a `DynSender<N>` provided `N` implements `Into<M>`
-impl<M: Message, N: Message + Into<M>> From<mpsc::UnboundedSender<M>> for DynSender<N> {
-    fn from(sender: mpsc::UnboundedSender<M>) -> Self {
-        Box::new(sender)
+/// An `mpsc::Sender<M>` is a `DynSender<M>`
+#[async_trait]
+impl<M: Message, N: Message + Into<M>> Sender<N> for mpsc::Sender<M> {
+    async fn send(&mut self, message: N) -> Result<(), ChannelError> {
+        Ok(SinkExt::send(&mut self, message.into()).await?)
     }
 }
 
+/// An `mpsc::UnboundedSender<M>` is a `DynSender<N>` provided `N` implements `Into<M>`
 #[async_trait]
 impl<M: Message, N: Message + Into<M>> Sender<N> for mpsc::UnboundedSender<M> {
     async fn send(&mut self, message: N) -> Result<(), ChannelError> {
         Ok(SinkExt::send(&mut self, message.into()).await?)
     }
-
-    fn sender_clone(&self) -> DynSender<N> {
-        Box::new(self.clone())
-    }
-
-    fn close_sender(&mut self) {
-        self.close_channel();
-    }
 }
 
-/// Make a `DynSender<N>` from a `DynSender<M>`
-///
-/// This is a workaround to the fact the compiler rejects a From implementation:
+/// A `oneshot::Sender<M>` is a `Sender<N>` provided `N` implements `Into<M>`
 ///
-/// ```shell
+/// There is one caveat. The `oneshot::Sender::send()` method consumes the sender,
+/// hence the one shot sender is wrapped inside an `Option`.
 ///
-///  impl<M: Message, N: Message + Into<M>> From<DynSender<M>> for DynSender<N> {
-///     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-///     |
-///     = note: conflicting implementation in crate `core`:
-///             - impl<T> From<T> for T;
-/// ```
-pub fn adapt<M: Message, N: Message + Into<M>>(sender: &DynSender<M>) -> DynSender<N> {
-    Box::new(sender.clone())
-}
-
+/// Such a [Sender] can only be used once:
+/// - it cannot be cloned
+/// - any message sent after a first one will be silently ignored
+/// - a message sent while the receiver has been drop will also be silently ignored
 #[async_trait]
-impl<M: Message, N: Message + Into<M>> Sender<N> for DynSender<M> {
+impl<M: Message, N: Message + Into<M>> Sender<N> for Option<oneshot::Sender<M>> {
     async fn send(&mut self, message: N) -> Result<(), ChannelError> {
```

Comment:
This implementation doesn't raise errors, but some `Sender` implementations do.

That said it would make sense for a one shot sender to raise an error if used twice.